### PR TITLE
Changed validateCollateralContainsNonADA so its clearer.

### PR DESCRIPTION
**Pull request by TimSheard**
Changed validateCollateralContainsNonADA, so the logic is easier to follow.

Small change to just 1 file. The logic for validateCollateralContainsNonADA was complicated and hard to follow,
because of nested case statements, if statements, failureUnless, and  logical conjunction.
Now it is broken into a Case statement with exactly 5 cases, and each case is clear.
This should be a non breaking change, as the function computes the same answer.

- [ x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 1]
```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
...
-  -- not just the Coin.
-  if areAllAdaOnly utxoCollateral && areAllAdaOnly (txBody ^. collateralReturnTxBodyL)
-    then pure ()
-    else failureUnless (Val.isAdaOnly bal) $
-      case txBody ^. collateralReturnTxBodyL of
-        SJust retTxOut
-          | not (Val.isAdaOnly colbal) ->
-              CollateralContainsNonADA (retTxOut ^. valueTxOutL)
-        _ -> CollateralContainsNonADA colbal
+  -- When we do not have any non-ada TxOuts we can short-circuit the more expensive
```
Comment (teodanciu) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:236]: "We do not have any non-ada TxOuts"  means we have ada-only TxOuts, correct?  Or maybe: we don't have multiassets?  Might be easier to understand without the double-negative, unless it's expressing something in particular

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 2]
```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
...
+  -- validation, which has to compute the full balance on the Values, not just the Coin.
+  -- By using 'areAllAdaOnly', the (True,True) case, this approach is both cheap and common,
+  -- and the computations of colInputsBal and totalBal, in the where clause, are not computed at all.
+  case (areAllAdaOnly utxoCollateral, areAllAdaOnly (txBody ^. collateralReturnTxBodyL)) of
+    (True, True) -> pure ()
+    (False, False) ->
+      if (Val.isAdaOnly totalBal)
+        then pure () -- Both have MultiSssets, and they cancel out.
+        else failureIf True $ CollateralContainsNonADA colInputBal
```
Comment (teodanciu) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:245]: I think we can use `failure $ CollateralContainsNonADA colInputBal` instead if we 
add `failure` to the imports above :
`import Validation (Validation, failure, failureIf, failureUnless,)`

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 3]
```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
...
+    (False, True) -> failureIf True $ CollateralContainsNonADA colInputBal
```
Comment (teodanciu) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:246]: Same as above: `failure $ CollateralContainsNonADA colInputBal`

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 4]
```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
...
+    (True, False) -> case txBody ^. collateralReturnTxBodyL of
+      SJust retTxOut -> failureIf True $ CollateralContainsNonADA (retTxOut ^. valueTxOutL)
```
Comment (teodanciu) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:248]: `failure $ CollateralContainsNonADA (retTxOut ^. valueTxOutL)`

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs [hunk 5]
```diff
@@ -233,24 +233,32 @@ validateCollateralContainsNonADA ::
...
-              CollateralContainsNonADA (retTxOut ^. valueTxOutL)
-        _ -> CollateralContainsNonADA colbal
+  -- When we do not have any non-ada TxOuts we can short-circuit the more expensive
+  -- validation, which has to compute the full balance on the Values, not just the Coin.
+  -- By using 'areAllAdaOnly', the (True,True) case, this approach is both cheap and common,
+  -- and the computations of colInputsBal and totalBal, in the where clause, are not computed at all.
+  case (areAllAdaOnly utxoCollateral, areAllAdaOnly (txBody ^. collateralReturnTxBodyL)) of
+    (True, True) -> pure ()
+    (False, False) ->
+      if (Val.isAdaOnly totalBal)
```
Comment (teodanciu) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:243]: Unnecessary parentheses

**Review by teodanciu — COMMENTED** [review]
I added some superficial remarks on the code. 

I think it's clearer than in master, and it made me understand the logic, but it seems that there's a new version of this that only triggers computations when validation is enabled: https://github.com/input-output-hk/cardano-ledger/pull/3574

**Comment by lehins** [comment]
Closing in favor of #3574
